CMSgov/design-system
# [WNMGDS-2539] Fix incompatibilities with React 18 types

## Summary

WNMGDS-2539

Fixes #2742 

When application teams try upgrading to the latest version of `@types/react-dom` (`v18.2.x`), TypeScript will throw errors when it encounters uses of the `Table`, `TableCell`, and `FilterChip` components. Currently we only use v17 of the `react-dom` types, so we don't see these errors. By upgrading our types and using the `create-react-app-typescript` example, we can reproduce those errors and fix them.

- Fixes errors generated by `Table` and `TableCell` components and updates the rest of the `Table*` components' type definitions to be more consistent.
- Fixes error generated by `FilterChip`.

## How to test

To see the errors:
1. Check out 38cec30e6
2. Run `yarn install && yarn build:core`
3. Delete `examples/create-react-app-typescript/node_modules`
4. Cd into `examples/create-react-app-typescript` and run `yarn install && yarn start`
5. There should be TypeScript errors in the console related to Table and FilterChip

To see them fixed:
1. Checkout bd3a8f8
2. Run `yarn install && yarn build:core`
3. Delete `examples/create-react-app-typescript/node_modules`
4. Cd into `examples/create-react-app-typescript` and run `yarn install && yarn start`
5. Errors should be gone

Note that this fixes the errors reported in the issue, but it sounds like there might be more after these are fixed that we haven't been able to reproduce yet.

## Checklist
